README.md
<p align="center">
  <img src="https://github.com/odomu/aliyunpan/blob/main/static/images/icon_256x256.png" alt="NebulaGraph Data Intelligence Suite(ngdi)">
</p>

<p align="center">
    <em>阿里云盘小白羊版 v3 修复版</em>
</p>

<p align="center">
<strong>该项目只是满足个人的需求进行修改，部分功能能用就行😉</strong><br>
<strong>没有过多的精力修改，很多功能不完善，更新非常慢👽</strong>
</p>

<p align="center">
<a href="https://github.com/odomu" target="_blank">
<img src="https://img.shields.io/badge/作者-@odomu-000000.svg?style=flat-square&logo=GitHub" />
</a>
<a href="https://github.com/odomu/aliyunpan/actions" target="_blank">
<img src="https://github.com/odomu/aliyunpan/workflows/Build/badge.svg" />
</a>
<a href="https://github.com/odomu/aliyunpan/releases" target="_blank">
<img src="https://img.shields.io/github/downloads/odomu/aliyunpan/total.svg?style=flat-square" />
</a>
<a href="https://github.com/odomu/aliyunpan/releases" target="_blank">
<img src="https://img.shields.io/github/release/odomu/aliyunpan.svg?style=flat-square" />
</a>
<a href="LICENSE" target="_blank">
<img src="https://img.shields.io/github/license/odomu/aliyunpan.svg?style=flat-square" />
</a>
<a href="https://t.me/+n1YbKE0JNo41MWRh" target="_blank">
<img src="https://img.shields.io/badge/Telegram-%E7%BE%A4%E7%BB%84-blue" />
</a>
</p>

# 项目说明
[阿里云盘小白羊版v3的延续版 by PingKuNet](https://github.com/PingKuNet/aliyunpan)
<br>
[原阿里云盘小白羊版v3 by aliyunpan](https://github.com/liupan1890/aliyunpan)

基于阿里云盘网页版开发的PC客户端，支持win7-11，macOS，linux

# 功能
- 根据阿里云盘Open平台api开发的网盘客户端，支持win7-11，macOS，linux
- 支持同时登录多个账号管理。
- 提供特有的文件夹树，方便快速操作。
- 在线播放网盘中各种格式的高清原画视频，并支持外挂字幕、音轨和播放速度调整，播放列表。
- 显示文件夹体积，支持文件夹和文件的混合排序（文件名/体积/时间）。
- 可以通过远程Aria2功能将文件直接下载到远程的VPS/NAS。
- 支持批量重命名大量文件和多层嵌套的文件夹。
- 可以快速复制文件，预览视频的雪碧图，并直接删除文件。
- 能够管理数万文件夹和数万文件，并一次性列出文件夹中的全部文件。
- 支持一次性上传/下载百万级的文件/文件夹。

# 请作者喝一杯咖啡
<p align="left">
  <img height="280" width="280" src="https://github.com/odomu/aliyunpan/assets/50035498/75efbbf3-36c0-4991-9dff-4ae319d09e7c" />
</p>

# 鸣谢
@jkqxl @iD2073 @ybbluesky 等为小白羊提供了大量的优化建议


CHANGELOG.md
# 2023/11/22

#### 更新日志

1. **MPV播放器功能增强（推荐）**
   - 支持播放列表
   - 支持记忆播放历史
   - 支持列表自动加载字幕文件
2. PotPlayer播放器功能增强
   - 支持播放列表
   - 不支持列表自动加载字幕文件
   - 不支持记忆播放历史
3. **优化列表加载速度**
4. **修复MPV在部分情况下记录历史错误的问题**
5. 资源网站功能重构
6. 网页播放器支持原画
7. 导入分享支持筛选
8. 移除github代理下载
9. 添加更新日志和赞赏码
10. 其他细节优化
11. 修复Bug
12. 破坏性改变【缓存路径】 alixby3 => aliyunxby
- %AppData%\Roaming\alixby3 => %AppData%\Roaming\aliyunxby
- 安装此版本需要重新进行配置，以此解决冲突问题
- 可以将旧版本文件名重命名为aliyunxby

#### 注意事项

- 更新和Star相关
- 字幕文件必须和播放列表文件目录同级
- 播放过程中不能退出软件，否则无法连续播放
- **需要从上个版本升级，后续不提供源码**

# 2023/8/27

1. 支持相册管理功能
   - 支持删除、修改、查看、添加相册
   - 支持相册的收藏、回收站功能
   - 支持上传、下载相册文件
   - 支持移出相册文件到其他盘
   - 支持相册创建快传链接
2. 移除无用代码
3. 修复bug

# 2023/8/23

1. 修复历史遗留问题
2. 修复文件根路径显示错误
3. 修复分享按钮显示错误
4. 修复文件操作后(移动、复制、删除)，目录树没有自动刷新的问题
5. 网页播放器支持显示**ass字幕文件的特效**

# 2023/8/14

#### 更新日志【请看完注意事项】

1. 支持同时显示和加载备份盘和资源盘
2. 支持资源盘和备份盘**跨盘操作**，文件可以跨盘转移
   - **后续版本备份盘将不能转移文件到资源盘，反之可以**
   - 如果文件被冻结则移动失败
   - 相关公告：https://www.yuque.com/aliyundrive/zpfszx/fitzlb1uyy0pv0iw
3. 支持资源盘**上传、下载、移动、分享、快传**
4. 支持资源盘全盘搜索功能
5. 支持资源盘使用插件功能
6. 支持显示资源盘容量详情
7. 支持资源盘和备份盘创建快传链接
8. 支持资源盘创建快捷方式
9. **支持资源盘标记功能**
10. **支持资源盘管理回收站**
11. **支持资源盘管理收藏夹**
12. 支持管理已经创建的快传链接
13. 支持内嵌打开资源网站，快捷导入分享链接
14. **重命名添加编号支持覆盖文件名**
15. 部分界面微调，左侧目录树根节点显示展开/收缩
16. 优化并且修复部分bug
17. **修复重命名多次操作的bug**
18. **修复drive_id获取错误**
19. **修复网盘空间显示错误**
20. 修复右键菜单文件复制移动错误
21. 修复文件下载显示错误
22. 修复OpenApi手机扫码
23. 下载界面支持范围选择
24. win添加arm64打包支持
25. 扫描重复文件智能选择
26. 放映室入口调整
27. 文件大小自动刷新错误

#### 注意：

1. **资源盘需要更新手机app**到最新版本才能显示【否则会资源盘和备份盘相同】
2. 需要分享功能请更新此版本【请先清理软件所有数据后使用】，**报错请重置数据后进行尝试**
3. 备份盘现在不能够分享文件【只能快传】，**只能转移到资源盘进行分享**
4. 快传可以分享任意格式【**有效期只有一天，并且只能被一个人保存，每天有创建快传数限制**】
5. ~~**因为官方的骚操作不得不更新**~~
6. 下载和播放限速请打开OpenApi功能，查看Wiki文档
   ![image](https://github.com/odomu/aliyunpan/assets/50035498/e2798697-bf0b-4d14-9114-d14279d74cad)

# 2023/6/28

1. 添加OpenApi支持，兼容旧版Api，加速视频播放和下载
   - **若只填写AccessToken，则每3小时过期后需重新获取**
   - **若填写RefreshToken和Oauth链接，AccessToken会自动刷新【点击任意文件预览自动刷新】**
   - 支持多账号填写不同AccessToken和RefreshToken
   - 手机扫码功能未测试，需要申请[OpenApi账号](https://survey.alibaba.com/apps/zhiliao/I9Dd1Nl89)
   - 推荐：采用AList获取[AccessToken和RefreshToken](https://alist.nn.ci/zh/guide/drivers/aliyundrive_open.html)
   - 注意：
      - **自动刷新AccessToken后请勿删除，防止多次请求Outh令牌链接**
      - 同一 IP 在 60 分钟内请求 10 次，会出现 **TooManyRequests** 🚫避免滥用，请勿滥用🚫
      - 🚫当一天的下载量过大时就会限速，尽量减少下载量（OpenApi也可能限速）🚫
        ![image](https://user-images.githubusercontent.com/50035498/235416335-67aab197-2759-405b-90c5-7aba3bb1daed.png)
3. 增加全局QPS速率限制，减少**Too Many Requests**发生的概率
4. 更新数据库，安装后需要重新登录
5. **重构下载相关代码，优化下载速度和性能**
6. 增加删除下载中任务时会自动删除临时文件
7. **增加下载可选线程数，OpenApi默认使用16线程**
8. 增加个人信息vip图标显示，用于提示vip是否过期
9. **优化自动加载同名字幕的加载速度**
10. 修复连续播放相关问题，支持自动继承播放倍速
   - 内置网页播放器**支持同专辑播放列表（不是文件列表）、记忆播放时间、画质选择、字幕选择、倍速播放**
      - **快捷键：z:重置倍数 x:-0.5倍数 c:+0.5倍数 enter: 全屏**
      - 注意：部分文件格式可能不支持播放
11. 修复网页播放器进度保存问题
12. **更新下载的referer和agent，可以跑满带宽**
13. 修复速度和下载进度条显示错误
14. 修复创建下载任务可能失败的问题
15. 修复MacOS任务栏图标过大的问题
16. **优化下载时自动重试机制**
17. 支持复制当前账号Cookies
18. 修复Vip图标显示错误
19. 升级Electron版本到21.4.4，Ctrl + Shift + F12打开DevTools
20. **升级内置Aira2版本到1.36.1，支持多个平台**
21. 支持鼠标控制后退，点击空白处取消选中
22. 批量重命名支持多次重复操作，不会关闭窗口
23. 文件多选支持取消已选中的文件
24. 支持鼠标控制后退，点击空白处取消选中
25. 导入转存功能支持仅选中的文件和文件夹
26. 更新OpenApi域名，优化刷新机制
27. 自定义播放器支持手动选择网盘内的字幕文件

#### 注意

1. 如果未启用OpenApi，下载可能会限速
2. 下载速度慢可以**适当调整**下载线程数，长时间大线程数可能会限速

# 2023/2/24

1.修复UserDeviceIllegality问题
2.修复下载功能
3.修复打包问题

# 2022/10/24

注：因上传下载功能尚未做完，本次仅同步代码不发布安装包

0. 恢复文件上传功能，支持单次上传百万个文件/文件夹，优化上传任务的数据库文件体积过大的问题
1. 优化sha1并发计算逻辑，增加文件sha1的缓存，提升上传大文件的性能
2. Add 上传前弹窗，设置重名冲突(删除/覆盖/自动重命名/不上传)
3. Add 单独的总上传速度限制设置
4. Add 上传文件时只上传可以秒传的文件的设置
5. Add 单独的上传文件时的并发执行数设置(最大50文件同时上传)
6. Add 左侧文件树文件夹对拖放上传的支持
7. Add 上传中列表里文件夹视图
8. Add 优先上传小文件(100M)的支持
9. Fix 按住Ctrl时，点击CheckBox和点击空白处，选中结果不一致的BUG
10. Fix 右键菜单-打开文件位置，没有选中和滚动到指定文件的BUG
11. Fix 断网时，重试上传任务，可能导致上传任务的断点续传进度丢失的BUG

# 2022/09/18

注：因上传下载功能尚未做完，本次仅同步代码不发布安装包

1. Fix v3.5.23alpha中的20余处小BUG
2. Fix 登录时遇到二次短信验证时不能继续登录的BUG
3. 完善批量重命名功能
4. 完善颜色标记功能
5. 优化文件列表加载显示逻辑，现在很优雅了
6. 优化文件名排序，支持中文数字排序，按win习惯英文在前中文在后
7. 优化文件夹树性能(全部文件夹列出速度加快3倍，20万文件夹不卡顿，树内存占用减少60%)
8. 恢复显示文件夹体积(可按大小排序)，优化计算文件夹体积的逻辑(速度加快计算量减少)
9. 底部增加网盘空间信息和文件夹内文件总数量
10. 增加复制文件名和复制目录树的功能
11. 适配更新vite3.1.2，更新全部package到最新版

# 2022/05/23

v3.5.23alpha 开发人员测试版本

1. Fix 偶发的登录后显示空白窗口的 BUG
2. Fix 登录太多账号时切换账号弹窗不能正确显示的 BUG
3. 优化长时间不用后需要重新登录的问题
4. Fix 偶发的下载中显示空白列表的 BUG
5. Fix 多次弹出升级提示窗口的 BUG
6. Fix 目录下超过 200 个文件夹不能显示的 BUG
7. 重制 APP 设置页面，分组设置，配置保存到 setting.config
8. 优化快捷键(现在很多操作都支持键盘快捷键)

--

v3.05.23.alpha 新增加的功能

文件恢复，放映室，网页版播放器，彻底删除菜单，文件属性菜单，所有列表支持右键菜单，
整盘高级搜索，文件快速筛选，右侧文件拖放移动，文件标记，文件夹快捷方式，我订阅的公众号，
重复文件清理，扫描大文件，扫描重复文件，扫描违规文件，扫描空文件夹，网盘相册间复制，
批量重命名

# 2022/04/14

1. 修正因阿里云盘官网升级导致的无法使用
2. 优化访问频次问题
3. 修正违禁视频详情BUG

# 2021/12/05

1. 修正网盘内文件路径过长时下载失败 BUG
2. 优化快捷键功能（帮助文档里有完整的快捷键说明）
3. 增加收藏夹的搜索功能
4. 优化后退按钮（以前是返回父文件夹，现在是后退，最多后退 20 步）
5. 增加优先下载小文件选项
6. 增加雪碧图里视频信息的显示和保存雪碧图按钮
7. 网盘页顶部路径默认隐藏需要在设置里勾选显示
8. 修正m3u8播放链接15分钟后失效的 BUG (现在4小时)
9. 修正之前部分版本代理设置被覆盖的 BUG
10. 增加自定义缓存位置功能
11. 增加帮助文档

# 2021/11/29

1. 修正 2.11.28 下载显示出错的 BUG
1. 修正登录时遇到二次验证导致点击登录按钮无反应的 BUG
2. 修正自 v2.11.07 开始不能上传体积为 0 的文件的 BUG
3. 修正 v2.11.16 上传文件时占用内存过多的问题
4. 11.16 里 60 文件同时上传会占用大量内存，11.28 里优化为只占用 400MB 以内的内存

5. 修正传输完自动关机触发时机不准确的 BUG
6. 修正上传时遇到没有访问权限的文件/文件夹时上传中队列卡住的 BUG
7. 增加视频文件洗码功能
8. 增加右侧文件拖动到左侧文件夹树（移动文件）功能
9. 优化远程 Aria2 下载功能并修复断线后自动重连
10. 优化本地 Aria2，无法连接时会尝试自动重启一次 Aria2 进程
11. 增加任务栏的下载中上传中进度提示(win/mac)
12. 底部状态栏显示总传输的预估剩余时间
13. 增加上传时跳过同名文件的设置项
14. 优化盘内文件搜索支持选择分类
15. 一些细节优化

# 2021/11/14

1. 增加创建分享、编辑分享、管理分享功能
2. 完善导入分享功能，支持部分导入和完整导入
3. 增加网盘内文件搜索功能
4. 增加下载时自动过滤违规文件的设置
5. 增加对禁止分享的文件的图标
6. 增加文件/文件夹置顶功能
7. 增加主题跟随系统
8. 增加下载上传完成后自动关机设置
9. 完善 mac 和 linux 下自定义播放软件功能
10. 恢复视频文件雪碧图，增加复制 M3U8 链接和原始视频链接功能
11. 修正 linux 下上传时会自动过滤软链接文件

# 2021/11/07

1. 优化一次性上传或下载大量文件时的界面卡顿（参阅 挑战一百万.md）
2. 创建日期文件夹支持编号了
3. 新增定时清理回收站功能
4. 新增我创建的分享列表功能(没做完)
5. 增加文件列表显示限制(减少加载中)
6. 恢复导入阿里云盘分享链接功能(仅全部导入)
7. 恢复新建文件功能
8. 恢复版本升级提示功能
9. 修正 windows 下载位置不能选择根目录的 BUG
10. 修正不能上传大于 97.6GB 的文件的问题
11. 修正上传速度显示不准确的 BUG
12. 修正不能同时下载同一个文件(下载到不同的位置)的 BUG
13. 优化部分文件格式图标

# 2021/10/31

1. 增加文件列表的缩略图模式

src/utils/aria2c.ts
import AliFile from '../aliapi/file'
import { IAliFileResp } from '../aliapi/dirfilelist'

import Aria2 from 'aria2-lib'
import axios from 'axios'
import DownDAL, { IAriaDownProgress, IStateDownFile } from '../down/DownDAL'
import message from './message'
import UserDAL from '../user/userdal'
import { useFootStore, useSettingStore } from '../store'
import DebugLog from './debuglog'
import Config from '../config'
import AliTrash from '../aliapi/trash'

import path from 'path'
import fs from 'fs'

const localPwd = 'S4znWTaZYQi3cpRNb'

let Aria2cChangeing: boolean = false
let Aria2EngineLocal: Aria2 | undefined = undefined
let Aria2EngineRemote: Aria2 | undefined = undefined

let IsAria2cOnlineLocal: boolean = false

let Aria2cLocalRelaunchTime = 0

let IsAria2cOnlineRemote: boolean = false

let Aria2cRemoteRetryTime = 0

function GetAria() {
  if (useSettingStore().AriaIsLocal) return Aria2EngineLocal
  return Aria2EngineRemote
}

function SetAriaOnline(isOnline: boolean, ariaState: string = '') {
  if (!ariaState) ariaState = useSettingStore().ariaState
  if (ariaState == 'local') {
    IsAria2cOnlineLocal = isOnline
    let ariaInfo = isOnline ? 'Aria ⚯ Local' : 'Aria 已断开'
    useFootStore().mSaveAriaInfo(ariaInfo)
  } else {
    IsAria2cOnlineRemote = isOnline
    let ariaInfo = isOnline ? 'Aria ⚯ ' + (Aria2EngineRemote?.host || '') : 'Aria 已断开'
    useFootStore().mSaveAriaInfo(ariaInfo)
  }
}

function CloseRemote() {
  if (IsAria2cOnlineRemote) {
    IsAria2cOnlineRemote = false
    if (Aria2EngineRemote) {
      try {
        Aria2EngineRemote.close().catch().then()
      } catch {}
      Aria2EngineRemote = undefined
    }
  }
}

export function IsAria2cRemote() {
  return IsAria2cOnlineRemote
}

export async function AriaTest(https: boolean, host: string, port: number, secret: string) {
  const url = (https ? 'https://' : 'http://') + host + ':' + port.toString() + '/jsonrpc'
  return axios
    .post(
      url,
      { method: 'aria2.getGlobalStat', jsonrpc: '2.0', id: 'id' + Date.now(), params: ['token:' + secret] },
      {
        responseType: 'json',
        headers: {
          'Content-Type': 'application/json'
        },
        timeout: 4000
      }
    )
    .then(() => {
      return true
    })
    .catch(function(error) {
      if (error.response && error.response.data && error.response.data.error) {
        if (error.response.data.error.message == 'Unauthorized') {
          message.error('连接失败 密码错误 ' + url + ' secret=' + secret)
          return false
        }
      }
      if (error.message && error.message.indexOf('timeout of') >= 0) {
        message.error('连接失败 网络连接超时 ' + url)
        return false
      }
      message.error('连接失败 ' + (error.message ? error.message : '') + ' ' + url + ' secret=' + secret)
      return false
    })
}

function Sleep(msTime: number) {
  return new Promise((resolve) =>
    setTimeout(
      () =>
        resolve({
          success: true,
          time: msTime
        }),
      msTime
    )
  )
}


export async function AriaChangeToRemote() {
  if (Aria2cChangeing) return undefined
  Aria2cChangeing = true
  CloseRemote()
  try {
    const settingStore = useSettingStore()
    const host = settingStore.ariaUrl.split(':')[0]
    const port = parseInt(settingStore.ariaUrl.split(':')[1])
    const secret = settingStore.ariaPwd

    const options = { host, port, secure: settingStore.ariaHttps, secret, path: '/jsonrpc' }
    Aria2EngineRemote = new Aria2({ WebSocket: global.WebSocket, fetch: window.fetch.bind(window), ...options })

    Aria2EngineRemote.on('close', () => {
      if (IsAria2cOnlineRemote && !Aria2cChangeing) {
        if (!settingStore.AriaIsLocal) {
          message.error('Aria2远程连接已断开')
          SetAriaOnline(false, 'remote')
        }
      }
    })
    await Sleep(500)
    await Aria2EngineRemote.open()
      .then(() => {
        Aria2cRemoteRetryTime = 0
        SetAriaOnline(true, 'remote')
      })
      .catch(() => {
        Aria2cRemoteRetryTime++
        SetAriaOnline(false, 'remote')
      })

    if (!IsAria2cOnlineRemote) {
      const url = host + ':' + port + ' secret=' + secret
      if (!settingStore.AriaIsLocal && Aria2cRemoteRetryTime % 10 == 1) message.error('无法连接到远程Aria2 ' + url)
    } else {
      await AriaGlobalSpeed()
    }
  } catch (e) {
    SetAriaOnline(false, 'remote')
  }
  Aria2cChangeing = false
  return IsAria2cOnlineRemote
}


export async function AriaChangeToLocal() {
  CloseRemote()
  if (Aria2cLocalRelaunchTime < 10) {
    try {
      let port = 16800
      if (Aria2EngineLocal == undefined) {
        port = window.WebRelaunchAria ? await window.WebRelaunchAria() : 16800
        const options = { host: '127.0.0.1', port, secure: false, secret: localPwd, path: '/jsonrpc' }
        Aria2EngineLocal = new Aria2({ WebSocket: global.WebSocket, fetch: window.fetch.bind(window), ...options })
        Aria2EngineLocal.on('close', () => {
          IsAria2cOnlineLocal = false
          if (useSettingStore().AriaIsLocal) {
            if (Aria2cLocalRelaunchTime < 2) {
              message.error('Aria2本地连接已断开')
            }
            SetAriaOnline(false, 'local')
          }
        })
        Aria2EngineLocal.setMaxListeners(0)
      }
      await Sleep(500)
      await Aria2EngineLocal.open()
        .then(() => {
          Aria2cLocalRelaunchTime = 0
          SetAriaOnline(true, 'local')
        })
        .catch(() => {
          SetAriaOnline(false, 'local')
          Aria2cLocalRelaunchTime++
          if (Aria2cLocalRelaunchTime < 2) {
            message.info('正在尝试重启Aria进程中。。。')
          }
        })

      if (!IsAria2cOnlineLocal) {
        const url = `127.0.0.1:${port} secret=${localPwd}`
        if (Aria2cLocalRelaunchTime < 2) message.error('无法连接到本地Aria2 ' + url)
      } else {
        await AriaGlobalSpeed()
      }
    } catch (e) {
      SetAriaOnline(false, 'local')
    }
  } else {
    Aria2EngineLocal = undefined
  }
  return true
}


export async function AriaGlobalSpeed() {
  try {
    const settingStore = useSettingStore()
    const limit = settingStore.downGlobalSpeed.toString() + (settingStore.downGlobalSpeedM == 'MB' ? 'M' : 'K')
    await GetAria()?.call('aria2.changeGlobalOption', { 'max-overall-download-limit': limit }).catch((e: any) => {
      if (e && e.message == 'Unauthorized') message.error('Aria2密码错误(密码不要有 ^ 或特殊字符)')
      IsAria2cOnlineLocal = false
    })
  } catch {
    SetAriaOnline(false)
  }
}

export async function AriaConnect() {
  if (useSettingStore().AriaIsLocal) {
    if (!IsAria2cOnlineLocal) await AriaChangeToLocal()
    return IsAria2cOnlineLocal
  } else {
    if (!IsAria2cOnlineRemote) await AriaChangeToRemote()
    return IsAria2cOnlineRemote
  }
}


export async function AriaGetDowningList() {
  const multicall = [
    ['aria2.tellActive', ['gid', 'status', 'totalLength', 'completedLength', 'downloadSpeed', 'errorCode', 'errorMessage']],
    ['aria2.tellWaiting', 0, 1000, ['gid', 'status', 'totalLength', 'completedLength', 'downloadSpeed', 'errorCode', 'errorMessage']],
    ['aria2.tellStopped', 0, 1000, ['gid', 'status', 'totalLength', 'completedLength', 'downloadSpeed', 'errorCode', 'errorMessage']]
  ]
  try {
    const result: any = await GetAria()?.multicall(multicall)
    if (result) {
      let list: IAriaDownProgress[] = []
      let arr = result[0][0]
      list = list.concat(arr)
      arr = result[1][0]
      list = list.concat(arr)
      arr = result[2][0]
      list = list.concat(arr)
      DownDAL.mSpeedEvent(list || [])
      SetAriaOnline(true)
    }
  } catch (e: any) {
    DebugLog.mSaveLog('danger', 'AriaGetDowningList' + (e.message || ''), e)
    SetAriaOnline(false)
  }
}


export async function AriaDeleteList(list: string[]) {
  const multicall = []
  for (let i = 0, maxi = list.length; i < maxi; i++) {
    multicall.push(['aria2.forceRemove', list[i]])
    multicall.push(['aria2.removeDownloadResult', list[i]])
  }
  try {
    await GetAria()?.multicall(multicall)
    SetAriaOnline(true)
  } catch {
    SetAriaOnline(false)
  }
}


export async function AriaStopList(list: string[]) {
  const multicall = []
  for (let i = 0, maxi = list.length; i < maxi; i++) {
    multicall.push(['aria2.forcePause', list[i]])
  }
  try {
    await GetAria()?.multicall(multicall)
    SetAriaOnline(true)
  } catch {
    SetAriaOnline(false)
  }
}


export function AriaShoutDown() {
  if (useSettingStore().AriaIsLocal) {
    Aria2EngineLocal?.call('aria2.forceShutdown').catch((e: any) => {})
  }
}


export async function AriaAddUrl(file: IStateDownFile): Promise<string> {
  try {
    const info = file.Info
    const token = UserDAL.GetUserToken(info.user_id)
    if (!token || !token.access_token) return '账号失效，操作取消'
    if (info.isDir) {
      const dirFull = path.join(info.DownSavePath, info.name)
      if (!info.ariaRemote) {
        try {
          await fs.promises.mkdir(dirFull, { recursive: true })
        } catch (error: any) {
          const errorMap: Record<string, string> = {
            EPERM: '文件没有读取权限',
            EBUSY: '文件被占用或锁定中',
            EACCES: '文件没有读取权限'
          }
          const errorMessage = errorMap[error.code] || error.message
          DebugLog.mSaveLog('danger', 'AriaAddUrl创建文件夹失败：' + dirFull + ' ' + (error || ''), error)
          return errorMessage
        }
      }
      const dirInfo: IAliFileResp = {
        items: [],
        itemsKey: new Set(),
        punished_file_count: 0,
        next_marker: '',
        m_user_id: info.user_id,
        m_drive_id: info.drive_id,
        dirID: info.file_id,
        dirName: info.name
      }
      do{
        const isGet = await AliTrash.ApiFileListOnePageAria('name', 'ASC', dirInfo)
        if (!isGet) return '解析子文件列表失败，稍后重试'
        if (file.Down.IsStop) {
          dirInfo.items.length = 0
          return '已暂停'
        }
        if (dirInfo.items.length > 0) {
          DownDAL.aAddDownload(dirInfo.items, dirFull, false)
          dirInfo.items.length = 0
        }
      } while (dirInfo.next_marker)
      return 'downed'
    } else {
      const dirPath = info.DownSavePath
      const outFileName = info.ariaRemote ? info.name : info.name + '.td'
      const fileFull = path.join(dirPath, info.name)
      if (!info.ariaRemote) {
        try {
          const fileStat = await fs.promises.stat(fileFull);
          if (fileStat && fileStat.size == info.size) return 'downed'
          else return '本地存在重名文件，请手动删除'
        } catch (error: any) {
          const errorMap: Record<string, string> = {
            EPERM: '文件没有读取权限',
            EBUSY: '文件被占用或锁定中',
            EACCES: '文件没有读取权限'
          }
          const errorMessage = errorMap[error.code] || error.message
          if (errorMessage.indexOf('no such file') < 0) {
            DebugLog.mSaveLog(
              'danger',
              `AriaAddUrl访问文件失败：${fileFull} ${errorMessage || ''}`,
              error
            )
            return errorMessage
          }
          if (info.size == 0) {
            try {
              await (await fs.promises.open(fileFull, 'w')).close()
              return 'downed'
            } catch {
              return '创建空文件失败'
            }
          }
        }
      }
      let downloadUrl = file.Down.DownUrl
      if (downloadUrl && downloadUrl.includes('x-oss-expires=')) {
        const expires = downloadUrl.split('x-oss-expires=')[1].split('&')[0]
        const lastTime = parseInt(expires) - Date.now() / 1000
        const needTime = (info.size + 1) / 1024 / 1024
        if (lastTime < 60 || lastTime < needTime + 60) {
          downloadUrl = ''
        }
      } else {
        downloadUrl = ''
      }
      if (!downloadUrl) {
        const durl = await AliFile.ApiFileDownloadUrl(info.user_id, info.drive_id, info.file_id, 14400)
        if (typeof durl == 'string') {
          return `生成下载链接失败, ${durl}`
        } else if (!durl.url) {
          DebugLog.mSaveLog('danger', `${info.file_id} 生成下载链接失败, ${JSON.stringify(durl)}`, null)
          return `生成下载链接失败,${JSON.stringify(durl)}`
        }
        downloadUrl = durl.url
        file.Down.DownUrl = downloadUrl
      }
      if (file.Down.IsStop) return '已暂停'
      const split = useSettingStore().downThreadMax
      const referer = Config.referer
      const userAgent = Config.downAgent
      const multicall = [
        ['aria2.forceRemove', info.GID],
        ['aria2.removeDownloadResult', info.GID],
        ['aria2.addUri', [downloadUrl], {
          gid: info.GID, dir: dirPath, out: outFileName,
          split, referer, 'user-agent': userAgent
        }]
      ]
      const result: any = await GetAria()?.multicall(multicall)
      if (result == undefined || result.length < 3
        || (result[2].code != undefined && result[2].code) != 0) {
        return '创建aria任务失败，稍后自动重试' + result[2].message
      }
      if (result[2].length == 1) return 'success'
    }
  } catch (e: any) {
    SetAriaOnline(false)
    DebugLog.mSaveLog('danger', 'AriaAddUrl' + (e.message || ''), e)
    return Promise.resolve('创建Aria任务失败连接断开')
  }
  return Promise.resolve('创建Aria任务失败1')
}


export function AriaHashFile(downitem: IStateDownFile): { DownID: string; Check: boolean } {
  const DownID = downitem.DownID
  const dir = downitem.Info.DownSavePath
  const out = downitem.Info.ariaRemote ? downitem.Info.name : downitem.Info.name + '.td'
  const sha1 = downitem.Info.sha1
  const crc64 = downitem.Info.crc64

  const data = {
    DownID: DownID,
    inputfile: path.join(dir, out),
    movetofile: path.join(dir, downitem.Info.name),
    hash: crc64 ? 'crc64' : sha1 ? 'sha1' : '',
    check: crc64 || sha1 || ''
  }
  let success = false
  if (data.inputfile == data.movetofile) {
    success = true
  } else {
    try {
      fs.renameSync(data.inputfile, data.movetofile)
      success = true
    } catch {
      try {
        fs.renameSync(data.inputfile, data.movetofile)
        success = true
      } catch (e: any) {
        DebugLog.mSaveLog('danger', 'AriaRename file=' + data.inputfile + ' error=' + (e.message || ''), e)
      }
    }
  }
  return { DownID, Check: success }
}


export function FormatAriaError(code: string, message: string): string {
  switch (code) {
    case '0':
      return ''
    case '1':
      return 'aria2c未知错误'
    case '2':
      return 'aria2c网络超时'
    case '3':
      return 'aria2c网络文件404'
    case '4':
      return 'aria2c网络文件404'
    case '5':
      return 'aria2c下载缓慢自动退出'
    case '6':
      return 'aria2c发生网络中断'
    case '7':
      return 'aria2c被强制退出错误'
    case '8':
      return 'aria2c服务器不支持断点续传'
    case '9':
      return 'aria2c本地硬盘空间不足'
    case '10':
      return 'aria2c分片大小更改'
    case '11':
      return 'aria2c重复任务'
    case '12':
      return 'aria2c重复BT任务'
    case '13':
      return 'aria2c文件已存在且不能覆盖'
    case '14':
      return 'aria2c文件重命名失败'
    case '15':
      return 'aria2c打开文件失败'
    case '16':
      return 'aria2c创建文件时失败'
    case '17':
      return 'aria2c文件写入失败'
    case '18':
      return 'aria2c创建文件夹失败, 请设置保存路径'
    case '19':
      return 'aria2cDNS解析失败'
    case '20':
      return 'aria2c解析磁力失败'
    case '21':
      return 'aria2cFTP不支持的命令'
    case '22':
      if (message.includes('403')) return '服务器拒绝访问403'
      if (message.includes('503')) return '服务器返回错误503'
      return message
    case '23':
      return 'aria2cHTTP重定向失败'
    case '24':
      return 'aria2cHTTP认证失败'
    case '25':
      return 'aria2c格式化种子失败'
    case '26':
      return 'aria2c读取种子信息失败'
    case '27':
      return 'aria2c磁力链接错误'
    case '28':
      return 'aria2c提供了错误的参数'
    case '29':
      return 'aria2c服务器超载暂时无法处理请求'
    case '30':
      return 'aria2cRPC传输参数错误'
    case '31':
      return 'aria2c多余的响应数据'
    case '32':
      return 'aria2c文件sha1校验失败'
    default:
      return message
  }
}
